util/src/main/scala/org/broadinstitute/dsde/rawls/util/Retry.scala

```diff
@@ -41,26 +53,35 @@ trait Retry {
    */
   def retryUntilSuccessOrTimeout[T](pred: Predicate[Throwable] = always, failureLogMessage: String = defaultErrorMessage)(interval: FiniteDuration, timeout: FiniteDuration)(op: () => Future[T])(implicit executionContext: ExecutionContext): Future[T] = {
     val trialCount = Math.ceil(timeout / interval).toInt
-    retry(Seq.fill(trialCount)(interval))(op, pred, failureLogMessage)
+    retryInternal(Seq.fill(trialCount)(interval), pred, failureLogMessage)(op)
   }
 
-  private def retry[T](remainingBackoffIntervals: Seq[FiniteDuration])(op: => () => Future[T], pred: Predicate[Throwable], failureLogMessage: String)(implicit executionContext: ExecutionContext): Future[T] = {
-    op().recoverWith {
-      case t if pred(t) && !remainingBackoffIntervals.isEmpty =>
-        logger.info(s"$failureLogMessage: ${remainingBackoffIntervals.size} retries remaining, retrying in ${remainingBackoffIntervals.head}", t)
-        after(remainingBackoffIntervals.head, system.scheduler) {
-          retry(remainingBackoffIntervals.tail)(op, pred, failureLogMessage)
-        }
-
-      case t =>
-        if (remainingBackoffIntervals.isEmpty) {
-          logger.info(s"$failureLogMessage: no retries remaining", t)
-        } else {
-          logger.info(s"$failureLogMessage: retries remain but predicate failed, not retrying", t)
-        }
-
-        Future.failed(t)
+  private def retryInternal[T](backoffIntervals: Seq[FiniteDuration],
+                               pred: Predicate[Throwable],
+                               failureLogMessage: String)
+                              (op: () => Future[T])
+                              (implicit executionContext: ExecutionContext): AccumulatingFuture[T] = {
+
+    def inner(remainingBackoffIntervales: Seq[FiniteDuration], errors: => List[Throwable]): AccumulatingFuture[T] = {
```

Comment:
¡intervales!

Maybe `def loop` instead? I made an inner function by habit to make the error accumulation tail-recursive. But it still might not be tail-recursive because of the `after(remainingBackoffIntervales.head, system.scheduler) { ... }` call. Will think about this more.